### PR TITLE
Jung/refactor

### PR DESCRIPTION
**Pull request by Frenchkebab**
- Unified terms `operator`, `generator` to `prover`.
- Replaced `jobId` to `bidId`
- Replaced `jobManager` to `proofMarketplace`
- access control for `setProverRewardShare`
- Created `Enum`, `Struct` library for better readability of the codebase

### CODE REVIEW

File: contracts/ProofMarketplace.sol [hunk 1]
```diff
@@ -742,42 +686,74 @@ contract ProofMarketplace is
...
+    function supportsInterface(
+        bytes4 interfaceId
+    ) public view virtual override(ERC165Upgradeable, AccessControlUpgradeable) returns (bool) {
+        return super.supportsInterface(interfaceId);
+    }
+
+    /**
+     @notice Enforces PMP to use only one matching engine image
+     */
+    function setMatchingEngineImage(bytes calldata pcrs) external onlyRole(UPDATER_ROLE) {
```
Comment (prateekreddy) [inline contracts/ProofMarketplace.sol:736]: `setMatchingEngineImage` and `verifyMatchingEngine` functions aren't really overrides instead they are matching engine management functions and have to segregated.

File: contracts/lib/Helper.sol [hunk 2]
```diff
@@ -89,8 +89,8 @@ library HELPER {
         return keccak256(abi.encode(roleId));
     }
 
-    function GENERATOR_FAMILY_ID(uint256 marketId) internal pure returns (bytes32) {
-        return keccak256(abi.encode("gen", marketId));
+    function PROVER_FAMILY_ID(uint256 marketId) internal pure returns (bytes32) {
+        return keccak256(abi.encode("prov", marketId));
```
Comment (prateekreddy) [inline contracts/lib/Helper.sol:93]: Please let @akshay111meher know about this.

File: contracts/staking/l2_contracts/SymbioticStaking.sol [hunk 3]
```diff
@@ -199,15 +182,15 @@ contract SymbioticStaking is
...
+        // update Vault and Prover stake amount
+        // update rewardPerToken for each vault and prover in SymbioticStakingReward
         _submitVaultSnapshot(_captureTimestamp, _vaultSnapshots);
 
         _updateTxCountInfo(_numOfTxs, _captureTimestamp, STAKE_SNAPSHOT_TYPE);
 
         // when all chunks of VaultSnapshots are submitted
         if (_index == _numOfTxs - 1) {
-            submissionStatus[_captureTimestamp][msg.sender] |= STAKE_SNAPSHOT_MASK;
+            submissionStatus[_captureTimestamp][msg.sender] = Enum.SubmissionStatus.STAKE_SNAPSHOT_DONE;
```
Comment (prateekreddy) [inline contracts/staking/l2_contracts/SymbioticStaking.sol:193]: Stake snapshot and slash snapshot being sequential isn't ideal. We probably have to allow them in any order.

Comment (prateekreddy) [reply contracts/staking/l2_contracts/SymbioticStaking.sol:193]: What if they aren't submitted in order?

File: contracts/staking/l2_contracts/SymbioticStaking.sol [hunk 4]
```diff
@@ -241,36 +223,34 @@ contract SymbioticStaking is
...
-        // there could be no operator slashed
-        if (_jobSlashed.length > 0) IStakingManager(stakingManager).onSlashResult(_jobSlashed);
-
-        // TODO: unlock the selfStake and reward it to the transmitter
-        emit SlashResultSubmitted(msg.sender, _index, _numOfTxs, _imageId, _slashResultData, _proof);
+        // there could be no prover slashed
+        if (_taskSlashed.length > 0) IStakingManager(stakingManager).onSlashResultSubmission(_taskSlashed);
 
         // when all chunks of Snapshots are submitted
         if (_index == _numOfTxs - 1) {
```
Comment (prateekreddy) [inline contracts/staking/l2_contracts/SymbioticStaking.sol:230]: There doesn't seem to be any checks on index

File: contracts/ProofMarketplace.sol [hunk 5]
```diff
@@ -105,66 +74,29 @@ contract ProofMarketplace is
...
-
-    //-------------------------------- Constants and Immutable start --------------------------------//
+    //-------------------------------- Constants and Immutable end --------------------------------//
 
     //-------------------------------- State variables start --------------------------------//
-    Market[] public marketData;
+    Struct.Market[] public marketData;
 
-    AskWithState[] public listOfAsk;
+    Struct.BidWithState[] public listOfBid;
```
[the remark this answers is not in the document]
Comment (Frenchkebab) [reply contracts/ProofMarketplace.sol:87]: Resolved

File: contracts/staking/l2_contracts/SymbioticStaking.sol (hunk 4 above)
Comment (Frenchkebab) [reply contracts/staking/l2_contracts/SymbioticStaking.sol:230]: It's checked `_checkValidity`.
Did you mean anything else?

File: contracts/staking/l2_contracts/SymbioticStaking.sol (hunk 3 above)
Comment (Frenchkebab) [reply contracts/staking/l2_contracts/SymbioticStaking.sol:193]: Resolved

**Review by prateekreddy — APPROVED** [review]

**Pull request merged**